### DISCUSSION TITLE
Lua GDExtension for Godot 4

### DISCUSSION THREAD

**Discussion by gilzoide** [gilzoide/godot-lua-pluginscript#52] [Announcements]
Hi everybody!

As you may know, the Godot 4 native extension API was reworked from the ground up, so that all Godot 3/GDNative plugins do not work on it without major refactoring.
With that said, I'd like to let you know that Lua PluginScript will not be refactored for Godot 4/GDExtension.

Instead, I have created a new project for Godot 4: [Lua GDExtension](https://github.com/gilzoide/lua-gdextension).
This new project has different design and implementation details, like the following:
1. Instead of Lua being usable only as scripts with a single shared state, there are now native classes for creating and interacting with Lua states.
   Lua as a language for scripts will be implemented at some point, but for now you may already create as many Lua states as you want and interact with them using GDScript.
2. When passing tables from Lua to Godot, they are now wrapped in a `LuaTable` class, instead of being transformed to `Dictionary` automatically. There will also be classes for Lua functions, threads (coroutines) and userdata.
3. `String` and `StringName` variants are automatically unwrapped to Lua strings, to avoid errors when calling Lua functions.
4. Instead of relying in LuaJIT and its Foreign Function Interface (the FFI), this new plugin uses Lua/C bindings.
   This makes it possible to use the latest version of PUC-Rio Lua, as well as other implementations like MoonScript.
5. Instead of binding all built-in Godot types separately, the new plugin binds the `Variant` type and makes dynamic calls.
   This provides a much simpler, smaller and cleaner implementation, as well as being more future-proof.
   I know, dynamic calls are supposed to be slower than method bind calls, but done is better than perfect.

It is quite early in development, but with the help of [godot-cpp](https://github.com/godotengine/godot-cpp) and [sol2](https://github.com/ThePhD/sol2), the implementation is going quite smoothly.
Of course, the new project is open for contributions, feedbacks, questions, just use the Discussions or Issues tab in the repo =D

That's it, thank you all, I hope you enjoy and I'll see you there!

gilzoide

**Comment by Frityet** [comment]
Thank you so much! I really needed this

**Comment by gilzoide** [comment]
If any of you want to try it out, there are now automated builds for the new project, so that all you need to do is downlod the zip and extract it into your own projects, no manual compiling required!
More information at https://github.com/gilzoide/lua-gdextension/discussions/15
